dali/operators/generic/transpose/transpose.h

```diff
@@ -16,37 +16,82 @@
 #define DALI_OPERATORS_GENERIC_TRANSPOSE_TRANSPOSE_H_
 
 #include <algorithm>
+#include <functional>
 #include <memory>
 #include <vector>
 
-#include "dali/pipeline/operator/operator.h"
+#include "dali/kernels/common/utils.h"
 #include "dali/operators/generic/transpose/cutt/cutt.h"
+#include "dali/pipeline/operator/operator.h"
 
 namespace dali {
 
 namespace detail {
 
-template <typename T>
-T Permute(const T& in, const std::vector<int>& permutation) {
-  T out = in;
-  for (size_t i = 0; i < permutation.size(); i++) {
-    auto idx = permutation[i];
-    out[i] = in[idx];
+template <typename ShapeT>
+inline void RowToColumnMajor(ShapeT *dims, int *perm, size_t len) {
+  std::reverse(dims, dims + len);
+  std::reverse(perm, perm + len);
+  for (size_t i = 0; i < len; ++i) perm[i] = len - 1 - perm[i];
+}
+
+// enough to represent batches of 4D
+constexpr int kStaticShapeElements = 5;
+using VecInt = SmallVector<int, kStaticShapeElements>;
+
+/**
+ * @brief Remove dimensions equal to 1 and adjust the shape & permutation
+ *
+ * Optionally convert from Row Major to Column Major compatible description (required by cuTT)
+ */
+template <typename ShapeT = int>
+void PrepareArguments(SmallVector<ShapeT, kStaticShapeElements> &shape, VecInt &perm,
+                      bool transpose = false) {
+  DALI_ENFORCE(shape.size() == perm.size());
+
+  // cuTT does not handle dimensions with size 1 so we remove them
+  // (H, W, 1) is equivalent to (H, W)
+  auto it_shape = shape.begin();
+  auto it_perm = perm.begin();
+  SmallVector<ShapeT, kStaticShapeElements> erased;
+  while (it_shape != shape.end()) {
+    if (*it_shape == 1) {
+      erased.push_back(*it_perm);
+      it_shape = shape.erase(it_shape);
+      it_perm = perm.erase(it_perm);
+    } else {
+      ++it_shape;
+      ++it_perm;
+    }
+  }
+  // when some permutation element is erased all elements positions after it should be decreased
+  // by one like it doesn't exist at all
+  // sort elements to erase in descending order so we avoid situations like
+  // erased(0, 2), perm(3, 1) -> perm(2, 0) while it should be (1, 0)
+  std::sort(erased.begin(), erased.end(), std::greater<int>());
+  for (auto &pos : erased) {
+    for (auto &elm : perm) {
+       if (elm > pos) {
+         --elm;
+       }
+    }
+  }
+  if (transpose) {
+    RowToColumnMajor(shape.data(), perm.data(), shape.size());
   }
-  return out;
 }
 
 }  // namespace detail
 
 
 template <typename Backend>
-class Transpose : public Operator<Backend> {
+class TransposeBase : public Operator<Backend> {
```

Comment:
I wouldn't change this name. It's obvious, that `Transpose` in a base for `TransposeCpu` and `TransposeGpu`. And requires less writing. I'd say, that you use this `SomethingBase` when you want to name like
```
class MyClassBase;
class MyClass : MyClassBase
```

Reply:
Done